File: BinarySearch.py
import time
from random import randint

def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1

def measure_search_time(arr, target):
    start_time = time.time()
    binary_search(arr, target)
    end_time = time.time()
    return end_time - start_time

# Simulate best, worst, and average-case scenarios
arr = sorted([randint(0, 1000000) for _ in range(1000000)])

# Best case: target is in the middle
print("Best case time:", measure_search_time(arr, arr[len(arr)//2]))

# Worst case: target is not present
print("Worst case time:", measure_search_time(arr, -1))

# Average case: target is somewhere random
print("Average case time:", measure_search_time(arr, arr[randint(0, len(arr) - 1)]))

File: BinarySearch.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <algorithm>

int binarySearch(const std::vector<int>& arr, int target) {
    int left = 0, right = arr.size() - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] == target) {
            return mid;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return -1;
}

void analyzeSearchPerformance() {
    std::vector<int> arr(1000000);
    for (int i = 0; i < 1000000; ++i) {
        arr[i] = i;
    }

    // Measure best-case scenario
    auto start = std::chrono::high_resolution_clock::now();
    binarySearch(arr, arr[arr.size() / 2]);
    auto end = std::chrono::high_resolution_clock::now();
    std::cout << "Best case time: "
              << std::chrono::duration_cast<std::chrono::microseconds>(end - start).count()
              << " microseconds" << std::endl;

    // Measure worst-case scenario (target not present)
    start = std::chrono::high_resolution_clock::now();
    binarySearch(arr, -1);
    end = std::chrono::high_resolution_clock::now();
    std::cout << "Worst case time: "
              << std::chrono::duration_cast<std::chrono::microseconds>(end - start).count()
              << " microseconds" << std::endl;

    // Measure average-case scenario (random target)
    start = std::chrono::high_resolution_clock::now();
    binarySearch(arr, arr[arr.size() / 3]);
    end = std::chrono::high_resolution_clock::now();
    std::cout << "Average case time: "
              << std::chrono::duration_cast<std::chrono::microseconds>(end - start).count()
              << " microseconds" << std::endl;
}

int main() {
    analyzeSearchPerformance();
    return 0;
}
